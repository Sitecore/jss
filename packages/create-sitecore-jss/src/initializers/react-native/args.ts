import { BaseArgs, ClientAppAnswer } from '../../common';

export type ReactNativeArgs = BaseArgs & Partial<Omit<ClientAppAnswer, 'fetchWith'>>;
