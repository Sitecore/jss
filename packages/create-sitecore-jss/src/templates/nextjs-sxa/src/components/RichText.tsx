import React from 'react';
import { Field, RichText as JssRichText } from '@sitecore-jss/sitecore-jss-nextjs';

interface Fields {
  Text: Field<string>;
}

export type RichTextProps = {
  params: { [key: string]: string };
  fields: Fields;
};

export const Default = (props: RichTextProps): JSX.Element => {
  const text = props.fields ? (
    <JssRichText field={props.fields.Text} />
  ) : (
    <span className="is-empty-hint">Rich text</span>
  );
  const id = props.params.RenderingIdentifier;

  return (
    <div
      className={`component rich-text ${props.params.styles.trimEnd()}`}
      id={id ? id : undefined}
    >
      <div className="component-content">{text}</div>
    </div>
  );
};
