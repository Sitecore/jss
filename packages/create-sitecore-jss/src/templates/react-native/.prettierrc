{
  "endOfLine": "crlf",
  "singleQuote": true,
  "trailingComma": "es5",
  "tabWidth": 2,
  "arrowParens": "always",
  "printWidth": 100
}
