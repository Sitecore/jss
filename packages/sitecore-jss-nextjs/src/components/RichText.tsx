import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { useRouter } from 'next/router';
import {
  RichText as ReactRichText,
  RichTextPropTypes,
  RichTextProps as ReactRichTextProps,
} from '@sitecore-jss/sitecore-jss-react';

export type RichTextProps = ReactRichTextProps & {
  /**
   * Selector which should be used in order to prefetch it and attach event listeners
   * @default 'a[href^="/"]'
   */
  internalLinksSelector?: string;
};

const prefetched: { [cacheKey: string]: boolean } = {};

export const RichText = (props: RichTextProps): JSX.Element => {
  const { internalLinksSelector = 'a[href^="/"]', ...rest } = props;
  const hasText = props.field && props.field.value;
  const isEditing = props.editable && props.field && props.field.editable;

  const router = useRouter();
  const richTextRef = useRef<HTMLElement>(null);

  useEffect(() => {
    // NOT IN EXPERIENCE EDITOR
    if (hasText && !isEditing) {
      initializeLinks();
    }
  });

  const routeHandler = (ev: MouseEvent) => {
    if (!ev.target) return;

    ev.preventDefault();

    const pathname = (ev.target as HTMLAnchorElement).pathname;

    router.push(pathname, pathname, { locale: false });
  };

  const initializeLinks = () => {
    const node = richTextRef.current;

    // selects all links that start with '/'
    const internalLinks = node && node.querySelectorAll<HTMLAnchorElement>(internalLinksSelector);

    if (!internalLinks || !internalLinks.length) return;

    internalLinks.forEach((link) => {
      if (!prefetched[link.pathname]) {
        router.prefetch(link.pathname, undefined, { locale: false });
        prefetched[link.pathname] = true;
      }

      link.removeEventListener('click', routeHandler, false);
      link.addEventListener('click', routeHandler, false);
    });
  };

  return <ReactRichText ref={richTextRef} {...rest} />;
};

RichText.propTypes = {
  internalLinksSelector: PropTypes.string,
  ...RichTextPropTypes,
};

RichText.defaultProps = {
  tag: 'div',
  editable: true,
};

RichText.displayName = 'NextRichText';
