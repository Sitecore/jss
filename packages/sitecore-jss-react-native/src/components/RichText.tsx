import React from 'react';
import HtmlView from 'react-native-htmlview';
import PropTypes from 'prop-types';

export interface RichTextProps {
  [htmlViewProps: string]: unknown;
  /** The rich text field data. */
  field: {
    value?: string;
    editable?: string;
  } | null;
}

export const RichText: React.SFC<RichTextProps> = ({ field, ...otherProps }) => {
  if (!field || (!field.editable && !field.value)) {
    return null;
  }

  const htmlValue = field.editable ? field.editable : field.value;

  return <HtmlView value={htmlValue as string} {...otherProps} />;
};

RichText.propTypes = {
  field: PropTypes.shape({
    value: PropTypes.string,
    editable: PropTypes.string,
  }),
};

RichText.displayName = 'RichText';
