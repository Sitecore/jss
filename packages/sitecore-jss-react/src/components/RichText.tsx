import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';

export interface RichTextProps {
  [htmlAttributes: string]: unknown;
  /** The rich text field data. */
  field?: {
    value?: string;
    editable?: string;
  };
  /**
   * The HTML element that will wrap the contents of the field.
   * @default <div />
   */
  tag?: string;
  /**
   * Can be used to explicitly disable inline editing.
   * If true and `field.editable` has a value, then `field.editable` will be processed and rendered as component output. If false, `field.editable` value will be ignored and not rendered.
   * @default true
   */
  editable?: boolean;
}

export const RichText: React.SFC<RichTextProps> = forwardRef(
  ({ field, tag, editable, ...otherProps }, ref) => {
    if (!field || (!field.editable && !field.value)) {
      return null;
    }

    const htmlProps = {
      dangerouslySetInnerHTML: {
        __html: field.editable && editable ? field.editable : field.value,
      },
      ref,
      ...otherProps,
    };

    return React.createElement(tag || 'div', htmlProps);
  }
);

export const RichTextPropTypes = {
  field: PropTypes.shape({
    value: PropTypes.string,
    editable: PropTypes.string,
  }),
  tag: PropTypes.string,
  editable: PropTypes.bool,
};

RichText.propTypes = RichTextPropTypes;

RichText.defaultProps = {
  tag: 'div',
  editable: true,
};

RichText.displayName = 'RichText';
