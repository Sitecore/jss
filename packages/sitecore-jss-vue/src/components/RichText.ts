import { h, defineComponent, PropType } from 'vue';

export const RichText = defineComponent({
  props: {
    /** The rich text field data. */
    field: {
      type: Object as PropType<{
        value?: string;
        editable?: string;
      }>,
      default() {
        return {} as {
          value?: string;
          editable?: string;
        };
      },
      required: true,
    },
    /**
     * The HTML element that will wrap the contents of the field.
     */
    tag: { type: String, default: 'div' },
    /**
     * Can be used to explicitly disable inline editing.
     * If true and `field.editable` has a value, then `field.editable` will
     * be processed and rendered as component output.
     * If false, `field.editable` value will be ignored and not rendered.
     */
    editable: { type: Boolean, default: true },
  },
  mounted() {
    this.bindRouteLinks();
  },
  methods: {
    /**
     * Click handler for links.
     * @param {MouseEvent} event - event emmited by clicking the link
     */
    routeHandler(event: MouseEvent): void {
      event.preventDefault();
      let target = event.target as HTMLAnchorElement;
      /**
       * If the target is not the anchor itself we set the target
       * to be the closest anchor parent element
       */
      if (!target.pathname) {
        target = target.closest('a') as HTMLAnchorElement;
      }

      const destination = target.hash ? `${target.pathname}${target.hash}` : target.pathname;

      this.$router.push(destination);
    },
    /**
     * Extracts anchor elements and adds a custom click event
     * listener to prevent page refresh.
     */
    bindRouteLinks() {
      const hasText = this.$props.field && this.$props.field?.value;
      const isEditing = this.$props.editable && this.$props.field?.editable;

      if (hasText && !isEditing) {
        // selects all links that start with '/'
        const internalLinks = this.$el.querySelectorAll('a[href^="/"]') as NodeListOf<
          HTMLAnchorElement
        >;

        // Remove old and add new click event listener
        internalLinks.forEach((link) => {
          link.removeEventListener('click', this.routeHandler, false);
          link.addEventListener('click', this.routeHandler, false);
        });
      }
    },
  },
  render() {
    const { field, tag, editable } = this.$props;
    if (!field || (!field.editable && !field.value)) {
      return null;
    }

    // this.$data should be passed along to the
    // `createElement` function in order to retain attributes and events
    // https://v3.vuejs.org/guide/render-function.html#render-functions
    const data = {
      ...this.$data,
      innerHTML: field.editable && editable ? field.editable : field.value,
    };

    return h(tag || 'div', data);
  },
});
