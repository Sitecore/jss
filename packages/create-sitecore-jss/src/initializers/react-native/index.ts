import path from 'path';
import { prompt } from 'inquirer';
import { prompts, ReactNativeAnswer } from './prompts';
import { Initializer } from '../../common/Initializer';
import { transform } from '../../common/steps';
import { ReactNativeArgs } from './args';

export default class ReactNativeInitializer implements Initializer {
  get isBase() {
    return true;
  }

  async init(args: ReactNativeArgs) {
    const defaults = args.yes
      ? {
          appName: 'sitecore-jss-react-native',
          hostName: 'sitecore-jss-react-native.dev.local',
          language: '',
        }
      : {};

    const answers = await prompt<ReactNativeAnswer>(prompts, { ...defaults, ...args });

    const mergedArgs = {
      ...args,
      ...answers,
    };

    const templatePath = path.resolve(__dirname, '../../templates/react-native');
    await transform(templatePath, mergedArgs, {
      filter: (filePath: string) =>
        !!mergedArgs.language || !filePath.endsWith('{{language}}.json'),
    });

    const response = {
      appName: answers.appName,
    };

    return response;
  }
}
