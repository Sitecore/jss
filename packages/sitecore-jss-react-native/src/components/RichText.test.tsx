import React from 'react';
import renderer from 'react-test-renderer';
import { StyleSheet } from 'react-native';
import { RichText } from './RichText';

describe('<RichText />', () => {
  test('should render nothing with missing field', () => {
    const field = null;
    const rendered = renderer.create(<RichText field={field} />);
    expect(rendered).toMatchSnapshot();
  });

  test('should render nothing with missing editable and value', () => {
    const props = {
      field: {},
    };
    const rendered = renderer.create(<RichText {...props} />);
    expect(rendered).toMatchSnapshot();
  });

  test('should render editable value if present', () => {
    const props = {
      field: {
        value: 'this value should not be rendered',
        editable: '<p>this value should be rendered &copy;</p>',
      },
    };
    const rendered = renderer.create(<RichText {...props} />);

    expect(rendered).toMatchSnapshot();
  });

  test('should render value if editable is undefined', () => {
    const props = {
      field: {
        value: 'this value should be rendered',
        editable: '',
      },
    };
    const rendered = renderer.create(<RichText {...props} />);

    expect(rendered).toMatchSnapshot();
  });

  it('should render other attributes with other props provided', () => {
    const field = {
      editable: 'value with <p>markup</p>',
    };
    const rawTextStyles = StyleSheet.create({
      style: {
        fontSize: 22,
      }
    });
    const markupStyles = StyleSheet.create({
      p: {
        fontSize: 16,
        marginBottom: 10,
      },
    });

    // 'style' prop will be applied to the html View (outer container).
    // 'stylesheet' is used by any html components created/parsed within the htmlview.
    // 'textComponentProps' is applied to any text strings created/parsed within the htmlview.
    // https://github.com/jsdf/react-native-htmlview#customizing-things-even-further
    const rendered = renderer.create(
      <RichText
        field={field}
        style={{ flex: 1 }}
        stylesheet={markupStyles}
        textComponentProps={rawTextStyles}
      />
    );
    expect(rendered).toMatchSnapshot();
  });
});
