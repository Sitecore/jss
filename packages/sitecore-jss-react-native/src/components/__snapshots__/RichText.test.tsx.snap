// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RichText /> should render editable value if present 1`] = `
<View>
  <Text
    onLongPress={null}
    onPress={null}
  >
    <Text
      style={
        Array [
          null,
          Object {},
        ]
      }
    >
      this value should be rendered Â©
    </Text>
  </Text>
</View>
`;

exports[`<RichText /> should render nothing with missing editable and value 1`] = `null`;

exports[`<RichText /> should render nothing with missing field 1`] = `null`;

exports[`<RichText /> should render other attributes with other props provided 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 22,
        },
        null,
      ]
    }
  >
    value with 
  </Text>
  <Text
    onLongPress={null}
    onPress={null}
    style={
      Object {
        "fontSize": 16,
        "marginBottom": 10,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 22,
          },
          Object {
            "fontSize": 16,
            "marginBottom": 10,
          },
        ]
      }
    >
      markup
    </Text>
  </Text>
</View>
`;

exports[`<RichText /> should render value if editable is undefined 1`] = `
<View>
  <Text
    style={
      Array [
        null,
        null,
      ]
    }
  >
    this value should be rendered
  </Text>
</View>
`;
